crates/biome_js_semantic/src/format_semantic_model.rs

```diff
@@ -0,0 +1,201 @@
+            f,
+            [
+                text("Scope {"),
+                group(&block_indent(&format_args![
+                    format_with(|f| {
+                        let range = std::format!("{:?}", self.range());
+                        write!(
+                            f,
+                            [
+                                text("id: "),
+                                self.id,
+                                text(" @ "),
+                                dynamic_text(range.as_str(), TextSize::default()),
+                                hard_line_break()
+                            ]
+                        )?;
+
+                        Ok(())
+                    }),
+                    text("children ["),
+                    group(&block_indent(&format_with(|f| {
+                        if children.is_empty() {
+                            return write!(f, [text("No children")]);
+                        }
+                        for child in &children {
+                            match child {
+                                ScopeOrBinding::Scope(scope) => write!(f, [scope])?,
+                                ScopeOrBinding::Binding(binding) => write!(f, [binding])?,
+                            }
+                        }
+                        Ok(())
+                    }))),
+                    text("]"),
+                ])),
+                text("}"),
+                hard_line_break(),
+            ]
+        )?;
+        Ok(())
+    }
+}
+
+impl Format<FormatSemanticModelContext> for Binding {
+    fn fmt(&self, f: &mut Formatter<FormatSemanticModelContext>) -> FormatResult<()> {
+        write!(
+            f,
+            [
+                text("Binding {"),
+                group(&block_indent(&format_with(|f| {
+                    let range = std::format!("{:?}", self.syntax().text_trimmed_range());
+                    write!(
+                        f,
+                        [
+                            text("id: "),
+                            self.id,
+                            text(" @ "),
+                            dynamic_text(range.as_str(), TextSize::default()),
+                            hard_line_break()
+                        ]
+                    )?;
+                    write!(f, [text("scope: "), self.scope().id, hard_line_break()])?;
+                    if let Some(ident) = self.syntax().clone().cast::<JsIdentifierBinding>() {
+                        let name = ident.name_token()?;
+                        write!(
+                            f,
+                            [
+                                text("ident: "),
+                                located_token_text(&name, name.text_trimmed_range()),
+                                hard_line_break()
+                            ]
+                        )?;
+                    }
+                    if let Some(ident) = self.syntax().clone().cast::<TsIdentifierBinding>() {
```

Comment:
What happens to self.syntax() if we can't cast it? Don't we lose information?

Reply:
Yeah I suppose. I'll see if I can do something about that.